File: top_interview_questions/2_medium/2_linked_list/add_two_numbers/solution.py
# https://leetcode.com/problems/add-two-numbers/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        sum_ = l1.val + l2.val
        carry = sum_ // 10
        result = ListNode(sum_ % 10)
        result_head = result

        l1 = l1.next
        l2 = l2.next

        while l1 or l2:
            sum_ = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            carry = sum_ // 10
            result.next = ListNode(sum_ % 10)

            if l1:
                l1 = l1.next

            if l2:
                l2 = l2.next

            result = result.next

        if carry:
            result.next = ListNode(carry)

        return result_head

File: top_interview_questions/2_medium/2_linked_list/add_two_numbers/solution.go
package main

import "fmt"

func main() {
	a := &ListNode{
		Val: 2,
		Next: &ListNode{
			Val:  4,
			Next: &ListNode{Val: 3},
		},
	}
	b := &ListNode{
		Val: 5,
		Next: &ListNode{
			Val:  6,
			Next: &ListNode{Val: 4},
		},
	}

	sum := addTwoNumbers(a, b)
	for sum != nil {
		fmt.Print(sum.Val, " ")
		sum = sum.Next
	}
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	sum := l1.Val + l2.Val
	carry := int(sum / 10)
	result := &ListNode{Val: sum % 10}
	resultHead := result

	l1 = l1.Next
	l2 = l2.Next

	for l1 != nil || l2 != nil {
		sum = carry

		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}

		carry = int(sum / 10)
		result.Next = &ListNode{Val: sum % 10}
		result = result.Next
	}

	if carry != 0 {
		result.Next = &ListNode{Val: carry}
	}

	return resultHead
}
